File: src/fireedge/src/client/features/OneApi/image.js
/* ------------------------------------------------------------------------- *
 * Copyright 2002-2024, OpenNebula Project, OpenNebula Systems               *
 *                                                                           *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may   *
 * not use this file except in compliance with the License. You may obtain   *
 * a copy of the License at                                                  *
 *                                                                           *
 * http://www.apache.org/licenses/LICENSE-2.0                                *
 *                                                                           *
 * Unless required by applicable law or agreed to in writing, software       *
 * distributed under the License is distributed on an "AS IS" BASIS,         *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  *
 * See the License for the specific language governing permissions and       *
 * limitations under the License.                                            *
 * ------------------------------------------------------------------------- */
import { Actions, Commands } from 'server/utils/constants/commands/image'
import {
  oneApi,
  ONE_RESOURCES,
  ONE_RESOURCES_POOL,
} from 'client/features/OneApi'
import { UpdateFromSocket } from 'client/features/OneApi/socket'
import http from 'client/utils/rest'
import {
  FilterFlag,
  Image,
  Permission,
  IMAGE_TYPES_STR,
  IMAGE_TYPES_FOR_FILES,
  IMAGE_TYPES_FOR_IMAGES,
  IMAGE_TYPES_FOR_BACKUPS,
} from 'client/constants'
import { getType } from 'client/models/Image'
import {
  updateResourceOnPool,
  removeResourceOnPool,
  updateTemplateOnResource,
} from 'client/features/OneApi/common'

const { IMAGE } = ONE_RESOURCES
const { IMAGE_POOL } = ONE_RESOURCES_POOL

const imageApi = oneApi.injectEndpoints({
  endpoints: (builder) => ({
    getAllImages: builder.query({
      /**
       * Retrieves information for all or part of the images in the pool.
       *
       * @param {object} params - Request params
       * @param {FilterFlag} [params.filter] - Filter flag
       * @param {number} [params.start] - Range start ID
       * @param {number} [params.end] - Range end ID
       * @returns {Image[]} List of images
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_POOL_INFO
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      transformResponse: (data) => {
        const imagesPool = data?.IMAGE_POOL?.IMAGE
          ? Array.isArray(data.IMAGE_POOL.IMAGE)
            ? data.IMAGE_POOL.IMAGE
            : [data.IMAGE_POOL.IMAGE]
          : []

        return imagesPool
      },
      providesTags: (images) =>
        images
          ? [
              ...images.map(({ ID }) => ({ type: IMAGE_POOL, id: `${ID}` })),
              IMAGE_POOL,
            ]
          : [IMAGE_POOL],
    }),
    getImages: builder.query({
      /**
       * Retrieves information for all or part of the images in the pool.
       *
       * @param {object} params - Request params
       * @param {FilterFlag} [params.filter] - Filter flag
       * @param {number} [params.start] - Range start ID
       * @param {number} [params.end] - Range end ID
       * @returns {Image[]} List of images
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_POOL_INFO
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      transformResponse: (data) => {
        const imagesPool = data?.IMAGE_POOL?.IMAGE
          ? Array.isArray(data.IMAGE_POOL.IMAGE)
            ? data.IMAGE_POOL.IMAGE
            : [data.IMAGE_POOL.IMAGE]
          : []

        const images = imagesPool.filter((image) =>
          IMAGE_TYPES_FOR_IMAGES.some(
            (imageType) => imageType === getType(image)
          )
        )

        return images
      },
      providesTags: (images) =>
        images
          ? [
              ...images.map(({ ID }) => ({ type: IMAGE_POOL, id: `${ID}` })),
              IMAGE_POOL,
            ]
          : [IMAGE_POOL],
    }),
    getFiles: builder.query({
      /**
       * Retrieves information for all or part of the images in the pool.
       *
       * @param {object} params - Request params
       * @param {FilterFlag} [params.filter] - Filter flag
       * @param {number} [params.start] - Range start ID
       * @param {number} [params.end] - Range end ID
       * @returns {Image[]} List of images
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_POOL_INFO
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      transformResponse: (data) => {
        const imagesPool = data?.IMAGE_POOL?.IMAGE
          ? Array.isArray(data.IMAGE_POOL.IMAGE)
            ? data.IMAGE_POOL.IMAGE
            : [data.IMAGE_POOL.IMAGE]
          : []

        const files = imagesPool.filter((image) =>
          IMAGE_TYPES_FOR_FILES.some(
            (imageType) => imageType === getType(image)
          )
        )

        return files
      },
      providesTags: (images) =>
        images
          ? [
              ...images.map(({ ID }) => ({ type: IMAGE_POOL, id: `${ID}` })),
              IMAGE_POOL,
            ]
          : [IMAGE_POOL],
    }),
    getBackups: builder.query({
      /**
       * Retrieves information for all or part of the images in the pool.
       *
       * @param {object} params - Request params
       * @param {FilterFlag} [params.filter] - Filter flag
       * @param {number} [params.start] - Range start ID
       * @param {number} [params.end] - Range end ID
       * @returns {Image[]} List of images
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_POOL_INFO
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      transformResponse: (data) => {
        const imagesPool = data?.IMAGE_POOL?.IMAGE
          ? Array.isArray(data.IMAGE_POOL.IMAGE)
            ? data.IMAGE_POOL.IMAGE
            : [data.IMAGE_POOL.IMAGE]
          : []

        const backups = imagesPool.filter((image) =>
          IMAGE_TYPES_FOR_BACKUPS.some(
            (imageType) => imageType === getType(image)
          )
        )

        return backups
      },
      providesTags: (images) =>
        images
          ? [
              ...images.map(({ ID }) => ({ type: IMAGE_POOL, id: `${ID}` })),
              IMAGE_POOL,
            ]
          : [IMAGE_POOL],
    }),
    getImage: builder.query({
      /**
       * Retrieves information for the image.
       *
       * @param {object} params - Request params
       * @param {string|number} params.id - Image id
       * @param {boolean} [params.decrypt] - Optional flag to decrypt contained secrets, valid only for admin
       * @returns {Image} Get Image identified by id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_INFO
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      transformResponse: (data) => data?.IMAGE ?? {},
      providesTags: (_, __, { id }) => [{ type: IMAGE, id }],
      onCacheEntryAdded: UpdateFromSocket({
        updateQueryData: (updateFn) =>
          imageApi.util.updateQueryData('getImages', undefined, updateFn),
        resource: 'IMAGE',
      }),
      async onQueryStarted({ id }, { dispatch, queryFulfilled }) {
        try {
          const { data: resourceFromQuery } = await queryFulfilled

          dispatch(
            imageApi.util.updateQueryData(
              'getImages',
              undefined,
              updateResourceOnPool({ id, resourceFromQuery })
            )
          )
        } catch {
          // if the query fails, we want to remove the resource from the pool
          dispatch(
            imageApi.util.updateQueryData(
              'getImages',
              undefined,
              removeResourceOnPool({ id })
            )
          )
        }
      },
    }),
    allocateImage: builder.mutation({
      /**
       * Allocates a new image in OpenNebula.
       *
       * @param {object} params - Request params
       * @param {string} params.template - A string containing the template of the image on syntax XML
       * @param {string} params.datastore - The datastore ID
       * @param {boolean} [params.capacity] - `true` to avoid checking datastore capacity
       * @returns {number} Image id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_ALLOCATE
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: [IMAGE_POOL],
    }),
    uploadImage: builder.mutation({
      /**
       * Upload image.
       *
       * @param {object} params - request params
       * @param {object} params.file - image file
       * @param {Function} params.uploadProcess - upload process function
       * @returns {number} Virtual machine id
       * @throws Fails when response isn't code 200
       */
      queryFn: async ({ file, uploadProcess }) => {
        try {
          const data = new FormData()
          data.append('files', file)
          const response = await http.request({
            url: '/api/image/upload',
            method: 'POST',
            data,
            onUploadProgress: uploadProcess,
          })

          return { data: response.data }
        } catch (axiosError) {
          const { response } = axiosError

          return { error: { status: response?.status, data: response?.data } }
        }
      },
    }),
    cloneImage: builder.mutation({
      /**
       * Clones an existing image.
       *
       * @param {object} params - Request params
       * @param {string} params.id - The id of the image to be cloned
       * @param {string} params.name - Name for the new image
       * @param {string|-1} [params.datastore] - The ID of the target datastore
       * @returns {number} The new image id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_CLONE
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: [IMAGE_POOL],
    }),
    removeImage: builder.mutation({
      /**
       * Deletes the given image from the pool.
       *
       * @param {object} params - Request params
       * @param {string} params.id - The object id
       * @returns {number} Image id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_DELETE
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: [IMAGE_POOL],
    }),
    enableImage: builder.mutation({
      /**
       * Enables an image.
       *
       * @param {string} id - Image id
       * @returns {number} Image id
       * @throws Fails when response isn't code 200
       */
      query: (id) => {
        const name = Actions.IMAGE_ENABLE
        const command = { name, ...Commands[name] }

        return { params: { id, enable: true }, command }
      },
      invalidatesTags: (_, __, id) => [{ type: IMAGE, id }, IMAGE_POOL],
    }),
    disableImage: builder.mutation({
      /**
       * Disables an image.
       *
       * @param {string} id - Image id
       * @returns {number} Image id
       * @throws Fails when response isn't code 200
       */
      query: (id) => {
        const name = Actions.IMAGE_ENABLE
        const command = { name, ...Commands[name] }

        return { params: { id, enable: false }, command }
      },
      invalidatesTags: (_, __, id) => [{ type: IMAGE, id }, IMAGE_POOL],
    }),
    persistentImage: builder.mutation({
      /**
       * Sets the Image as persistent or not persistent.
       *
       * @param {number|string} params - Request params
       * @param {string} params.id - Image id
       * @param {boolean} params.persistent - `True` for persistent
       * @returns {number} Image id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_PERSISTENT
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [{ type: IMAGE, id }, IMAGE_POOL],
    }),
    changeImageType: builder.mutation({
      /**
       * Changes the type of an Image.
       *
       * @param {number|string} params - Request params
       * @param {string} params.id - Image id
       * @param {IMAGE_TYPES_STR} params.type - New type for the Image
       * @returns {number} Image id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_CHTYPE
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, id) => [{ type: IMAGE, id }, IMAGE_POOL],
    }),
    updateImage: builder.mutation({
      /**
       * Replaces the image template contents.
       *
       * @param {number|string} params - Request params
       * @param {string} params.id - Image id
       * @param {string} params.template - The new template contents on syntax XML
       * @returns {number} Image id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_UPDATE
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [
        { type: IMAGE, id },
        { type: IMAGE_POOL, id },
      ],
      async onQueryStarted(params, { dispatch, queryFulfilled }) {
        try {
          const patchImage = dispatch(
            imageApi.util.updateQueryData(
              'getImage',
              { id: params.id },
              updateTemplateOnResource(params)
            )
          )

          const patchImages = dispatch(
            imageApi.util.updateQueryData(
              'getImages',
              undefined,
              updateTemplateOnResource(params)
            )
          )

          queryFulfilled.catch(() => {
            patchImage.undo()
            patchImages.undo()
          })
        } catch {}
      },
    }),
    changeImagePermissions: builder.mutation({
      /**
       * Changes the permission bits of a Image.
       * If set any permission to -1, it's not changed.
       *
       * @param {object} params - Request parameters
       * @param {string|number} params.id - Image id
       * @param {Permission|'-1'} params.ownerUse - User use
       * @param {Permission|'-1'} params.ownerManage - User manage
       * @param {Permission|'-1'} params.ownerAdmin - User administrator
       * @param {Permission|'-1'} params.groupUse - Group use
       * @param {Permission|'-1'} params.groupManage - Group manage
       * @param {Permission|'-1'} params.groupAdmin - Group administrator
       * @param {Permission|'-1'} params.otherUse - Other use
       * @param {Permission|'-1'} params.otherManage - Other manage
       * @param {Permission|'-1'} params.otherAdmin - Other administrator
       * @returns {number} Image id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_CHMOD
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [{ type: IMAGE, id }],
    }),
    changeImageOwnership: builder.mutation({
      /**
       * Changes the ownership of a Image.
       * If set `user` or `group` to -1, it's not changed.
       *
       * @param {object} params - Request parameters
       * @param {string|number} params.id - Image id
       * @param {string|number|'-1'} [params.userId] - User id
       * @param {Permission|'-1'} [params.groupId] - Group id
       * @returns {number} Image id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_CHOWN
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [{ type: IMAGE, id }, IMAGE_POOL],
    }),
    renameImage: builder.mutation({
      /**
       * Renames a Image.
       *
       * @param {object} params - Request parameters
       * @param {string} params.id - Image id
       * @param {string} params.name - The new name
       * @returns {number} Image id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_RENAME
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [{ type: IMAGE, id }, IMAGE_POOL],
    }),
    deleteImageSnapshot: builder.mutation({
      /**
       * Deletes a snapshot from the Image.
       *
       * @param {object} params - Request parameters
       * @param {string} params.id - Image id
       * @param {string} params.snapshot - ID of the snapshot to delete
       * @returns {number} Snapshot ID
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_SNAPDEL
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [{ type: IMAGE, id }],
    }),
    revertImageSnapshot: builder.mutation({
      /**
       * Reverts image state to a previous snapshot.
       *
       * @param {object} params - Request parameters
       * @param {string} params.id - Image id
       * @param {string} params.snapshot - ID of the snapshot to revert to
       * @returns {number} Snapshot ID
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_SNAPREV
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [{ type: IMAGE, id }],
    }),
    flattenImageSnapshot: builder.mutation({
      /**
       * Flatten the snapshot of image and discards others.
       *
       * @param {object} params - Request parameters
       * @param {string} params.id - Image id
       * @param {string} params.snapshot - ID of the snapshot to revert to
       * @returns {number} Snapshot ID
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_SNAPFLAT
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [{ type: IMAGE, id }],
    }),
    lockImage: builder.mutation({
      /**
       * Locks an image. Lock certain actions depending on blocking level.
       * - `USE` (1): locks Admin, Manage and Use actions.
       * - `MANAGE` (2): locks Manage and Use actions.
       * - `ADMIN` (3): locks only Admin actions.
       * - `ALL` (4): locks all actions.
       *
       * @param {object} params - Request parameters
       * @param {string|number} params.id - Image id
       * @param {'1'|'2'|'3'|'4'} params.lock - Lock level
       * @returns {number} Image id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_LOCK
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [{ type: IMAGE, id }, IMAGE_POOL],
    }),
    unlockImage: builder.mutation({
      /**
       * Unlocks an image.
       *
       * @param {object} params - Request parameters
       * @param {string|number} params.id - Image id
       * @returns {number} Image id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_UNLOCK
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [{ type: IMAGE, id }, IMAGE_POOL],
    }),
    restoreBackup: builder.mutation({
      /**
       * Restores an image.
       *
       * @param {number|string} params - Request params
       * @param {string} params.id - Image id
       * @param {number} params.datastore - New type for the Image
       * @param {string} params.options - New type for the Image
       * @returns {number} Image id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.IMAGE_RESTORE
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, id) => [{ type: IMAGE, id }],
    }),
  }),
})

export const {
  // Queries
  useGetAllImagesQuery,
  useLazyGetAllImagesQuery,
  useGetImageQuery,
  useLazyGetImageQuery,
  useGetImagesQuery,
  useLazyGetImagesQuery,
  useGetFilesQuery,
  useLazyGetFilesQuery,
  useGetBackupsQuery,
  useLazyGetBackupsQuery,

  // Mutations
  useAllocateImageMutation,
  useCloneImageMutation,
  useRemoveImageMutation,
  useEnableImageMutation,
  useDisableImageMutation,
  usePersistentImageMutation,
  useChangeImageTypeMutation,
  useUpdateImageMutation,
  useChangeImagePermissionsMutation,
  useChangeImageOwnershipMutation,
  useRenameImageMutation,
  useDeleteImageSnapshotMutation,
  useRevertImageSnapshotMutation,
  useFlattenImageSnapshotMutation,
  useLockImageMutation,
  useUnlockImageMutation,
  useUploadImageMutation,
  useRestoreBackupMutation,
} = imageApi

export default imageApi

File: src/ods/lib/helpers/one/image.rb
# -------------------------------------------------------------------------- #
# Copyright 2002-2026, OpenNebula Project, OpenNebula Systems                #
#                                                                            #
# Licensed under the Apache License, Version 2.0 (the "License"); you may    #
# not use this file except in compliance with the License. You may obtain    #
# a copy of the License at                                                   #
#                                                                            #
# http://www.apache.org/licenses/LICENSE-2.0                                 #
#                                                                            #
# Unless required by applicable law or agreed to in writing, software        #
# distributed under the License is distributed on an "AS IS" BASIS,          #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   #
# See the License for the specific language governing permissions and        #
# limitations under the License.                                             #
#--------------------------------------------------------------------------- #

module OpenNebula

    module DocumentServer

        # Defines methods to manage resources in OpenNebula using the OCA API
        module OneHelper

            # Defines methods to manage Images in OpenNebula
            module Image

                RESOURCE_TYPE = 'image'
                WAIT_DELETE   = true

                # Creates an image in a datastore.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param template [Hash] Image template.
                # @param datastore_id [Integer] Destination datastore ID.
                # @param no_check_capacity [Boolean] Whether to bypass capacity checks.
                # @return [OpenNebula::Image, OpenNebula::Error] created image or an API error.
                def self.create(client, template, datastore_id, no_check_capacity: false)
                    template = Hash.to_raw(template)
                    return template if OpenNebula.is_error?(template)

                    return OpenNebula::Error.new(
                        'Image template cannot be empty', OpenNebula::Error::EACTION
                    ) if template.to_s.empty?

                    return OpenNebula::Error.new(
                        'Datastore ID cannot be nil', OpenNebula::Error::EACTION
                    ) if datastore_id.nil?

                    image = OpenNebula::Image.new(OpenNebula::Image.build_xml, client)

                    rc = image.allocate(template, datastore_id, no_check_capacity)
                    return rc if OpenNebula.is_error?(rc)

                    rc = image.info
                    return rc if OpenNebula.is_error?(rc)

                    image
                end

                # Returns an image body using symbolized keys.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param image_id [Integer] Image ID.
                # @param downcase [Boolean] Whether to downcase keys.
                # @return [Hash, OpenNebula::Error] image body or an API error.
                def self.body(client, image_id, downcase: true)
                    image = get(client, image_id)
                    return image if OpenNebula.is_error?(image)

                    body = image.to_hash['IMAGE']

                    return OpenNebula::Error.new(
                        "Cannot retrieve Image body for resource '#{image_id}'",
                        OpenNebula::Error::EACTION
                    ) unless body

                    body.deep_symbolize_keys(:downcase => downcase)
                end

                # Checks whether an image with a name exists.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param name [String] Image name.
                # @return [Boolean, OpenNebula::Error] existence result or an API error.
                def self.exists?(client, name)
                    image = find(client, name)
                    return image if OpenNebula.is_error?(image)

                    !image.nil?
                end

                # Checks whether an image ID exists in the pool.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param image_id [Integer] Image ID.
                # @return [Boolean, OpenNebula::Error] existence result or an API error.
                def self.exists_id?(client, image_id)
                    image_pool = OpenNebula::ImagePool.new(client, -1)

                    rc = image_pool.info
                    return rc if OpenNebula.is_error?(rc)

                    image_pool.any? {|image| image.id.to_i == image_id.to_i }
                end

                # Retrieves an image with its current information.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param image_id [Integer] Image ID.
                # @return [OpenNebula::Image, OpenNebula::Error] image or an API error.
                def self.get(client, image_id)
                    return OpenNebula::Error.new(
                        'Image ID cannot be nil', OpenNebula::Error::EACTION
                    ) if image_id.nil?

                    image = OpenNebula::Image.new_with_id(image_id, client)

                    rc = image.info
                    return rc if OpenNebula.is_error?(rc)

                    image
                end

                # Returns an image name.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param image_id [Integer] Image ID.
                # @return [String, OpenNebula::Error] image name or an API error.
                def self.name(client, image_id)
                    image = get(client, image_id)
                    return image if OpenNebula.is_error?(image)

                    name = image.name
                    return OpenNebula::Error.new(
                        "Cannot retrieve name for Image '#{image_id}'",
                        OpenNebula::Error::EACTION
                    ) if name.to_s.empty?

                    name
                end

                # Finds an image by name and optional datastore ID.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param name [String] Image name.
                # @param ds_id [Integer, nil] Datastore filter.
                # @return [OpenNebula::Image, nil, OpenNebula::Error] image, none, or an API error.
                def self.find(client, name, ds_id: nil)
                    image_pool = OpenNebula::ImagePool.new(client, -1)

                    rc = image_pool.info
                    return rc if OpenNebula.is_error?(rc)

                    image = image_pool.find do |i|
                        i.name == name && (ds_id.nil? || i['DATASTORE_ID'].to_s == ds_id.to_s)
                    end
                    return if image.nil?

                    rc = image.info
                    return rc if OpenNebula.is_error?(rc)

                    image
                end

                # Finds an image by a template attribute and optional datastore ID.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param attr [String] Template attribute path.
                # @param value [Object] Expected attribute value.
                # @param ds_id [Integer, nil] Datastore filter.
                # @return [OpenNebula::Image, nil, OpenNebula::Error] image, none, or an API error.
                def self.find_by_attr(client, attr, value, ds_id: nil)
                    image_pool = OpenNebula::ImagePool.new(client, -1)

                    rc = image_pool.info
                    return rc if OpenNebula.is_error?(rc)

                    image = image_pool.find do |i|
                        i["TEMPLATE/#{attr}"].to_s == value.to_s &&
                            (ds_id.nil? || i['DATASTORE_ID'].to_s == ds_id.to_s)
                    end
                    return if image.nil?

                    rc = image.info
                    return rc if OpenNebula.is_error?(rc)

                    image
                end

                # Finds the first image imported from a Marketplace appliance UUID.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param appliance_uuid [String] Marketplace appliance UUID.
                # @param ds_id [Integer] Datastore ID.
                # @return [OpenNebula::Image, nil, OpenNebula::Error] image, none, or an API error.
                def self.find_by_marketplace_uuid(client, appliance_uuid, ds_id)
                    return OpenNebula::Error.new(
                        'Marketplace appliance UUID cannot be empty',
                        OpenNebula::Error::EACTION
                    ) if appliance_uuid.to_s.empty?

                    return OpenNebula::Error.new(
                        'Datastore ID cannot be nil',
                        OpenNebula::Error::EACTION
                    ) if ds_id.nil?

                    image_pool = OpenNebula::ImagePool.new(client, -1)

                    rc = image_pool.info
                    return rc if OpenNebula.is_error?(rc)

                    images = []
                    image_pool.sort_by {|image| image.id.to_i }.each do |image|
                        next unless image['DATASTORE_ID'].to_s == ds_id.to_s

                        rc = image.info
                        return rc if OpenNebula.is_error?(rc)

                        path = image['PATH'].to_s
                        images << image if path.include?("/appliance/#{appliance_uuid}/")
                    end

                    return if images.empty?

                    if images.size > 1
                        Log.warn(
                            'ONE',
                            'Multiple appliance images found for appliance UUID ' \
                            "#{appliance_uuid} in datastore #{ds_id}. " \
                            "Using image #{images.first.id}."
                        )
                    end

                    images.first
                end

                # Updates or appends template content to an image.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param image_id [Integer] Image ID.
                # @param content [Hash] Template data.
                # @param append [Boolean] Whether to append the data.
                # @return [OpenNebula::Image, OpenNebula::Error] updated image or an API error.
                def self.update(client, image_id, content, append: false)
                    return OpenNebula::Error.new(
                        'Image ID cannot be nil', OpenNebula::Error::EACTION
                    ) if image_id.nil?

                    content = Hash.to_raw(content)
                    return content if OpenNebula.is_error?(content)

                    return OpenNebula::Error.new(
                        'Image content cannot be empty', OpenNebula::Error::EACTION
                    ) if content.to_s.empty?

                    image = OpenNebula::Image.new_with_id(image_id, client)

                    rc = image.update(content, append)
                    return rc if OpenNebula.is_error?(rc)

                    rc = image.info
                    return rc if OpenNebula.is_error?(rc)

                    image
                end

                # Deletes an image, optionally forcing and waiting for its removal.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param image_id [Integer] Image ID.
                # @param force [Boolean] Whether to force deletion.
                # @param wait [Boolean] Whether to wait for deletion.
                # @return [true, OpenNebula::Error] success or an API error.
                def self.delete(client, image_id, force: false, wait: false)
                    return OpenNebula::Error.new(
                        'Image ID cannot be nil', OpenNebula::Error::EACTION
                    ) if image_id.nil?

                    image = if wait
                                get(client, image_id)
                            else
                                OpenNebula::Image.new_with_id(image_id, client)
                            end
                    return image if OpenNebula.is_error?(image)

                    rc = image.delete(force)
                    return rc if OpenNebula.is_error?(rc)
                    return true unless wait

                    Resource.wait_until_deleted(image)
                end

            end

        end

    end

end
